libraries/SITL/SIM_GPS.cpp

```diff
@@ -1014,6 +1017,300 @@ uint32_t GPS::CalculateBlockCRC32(uint32_t length, uint8_t *buffer, uint32_t crc
+    const uint8_t PACKET_TYPE = 0x40; // Report Packet 40h (GENOUT)
+
+    // Before writing the GSOF data buffer, the GSOF header needs added between the DCOL header and the payload data frame.
+    // https://receiverhelp.trimble.com/oem-gnss/index.html#GSOFmessages_GSOF.html?TocPath=Output%2520Messages%257CGSOF%2520Messages%257C_____2
+
+    static uint8_t TRANSMISSION_NUMBER = 0; // Functionally, this is a sequence number
+    // Most messages, even GSOF49, only take one page. For SIM, assume it.
+    assert(size < 0xFA); // GPS SIM doesn't yet support paging
+    constexpr uint8_t PAGE_INDEX = 0; 
+    constexpr uint8_t MAX_PAGE_INDEX = 0;
+    const uint8_t gsof_header[3] = {
+        TRANSMISSION_NUMBER,
+        PAGE_INDEX,
+        MAX_PAGE_INDEX,
+
+    };
+    ++TRANSMISSION_NUMBER;
+
+    // A captured GSOF49 packet from BD940  has LENGTH field set to 0x6d = 109 bytes.
+    // A captured GSOF49 packet from BD940  has total bytes of 115 bytes.
+    // Thus, the following 5 bytes are not counted.
+    // 1) STX
+    // 2) STATUS
+    // 3) PACKET TYPE
+    // 4) LENGTH
+    // 5) CHECKSUM
+    // 6) ETX
+    // This aligns with manual's idea of data bytes:
+    // "Each message begins with a 4-byte header, followed by the bytes of data in each packet. The packet ends with a 2-byte trailer."
+    // Thus, for this implementation with single-page single-record per DCOL packet,
+    // the length is simply the sum of data packet size, the gsof_header size.    
+    const uint8_t length = size + sizeof(gsof_header);
+    const uint8_t dcol_header[4] {
+        STX,
+        STATUS,
+        PACKET_TYPE,
+        length
+    };
+
+
+
+    // Sum bytes (status + type + length + data bytes) and modulo 256 the summation
+    // Because it's a uint8, use natural overflow
+    uint8_t csum = STATUS + PACKET_TYPE + length;
+    for (size_t i = 0; i < ARRAY_SIZE(gsof_header); i++) {
+        csum += gsof_header[i];
+    }
+    for (size_t i = 0; i < size; i++) {
+        csum += buf[i];
+    }
+
+    constexpr uint8_t ETX = 0x03;
+    const uint8_t dcol_trailer[2] = {
+        csum,
+        ETX
+    };
+
+    write_to_autopilot((char*)dcol_header, sizeof(dcol_header));
+    write_to_autopilot((char*)gsof_header, sizeof(gsof_header));
+    write_to_autopilot((char*)buf, size);
+    write_to_autopilot((char*)dcol_trailer, sizeof(dcol_trailer));
+    const uint8_t total_size = sizeof(dcol_header) + sizeof(gsof_header) + size + sizeof(dcol_trailer);
+     // Validate length based on everything but DCOL h
+    if(dcol_header[3] != total_size - (sizeof(dcol_header) +  sizeof(dcol_trailer))) {
+        INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+    }
+}
+
+uint64_t GPS::pack_double_into_gsof_packet(const double& src)
+{
+    uint64_t dst;
+    static_assert(sizeof(src) == sizeof(dst));
+    memcpy(&dst, &src, sizeof(dst));
+    dst = htobe64(dst);
+    return dst;
+}
+
+uint32_t GPS::pack_float_into_gsof_packet(const float& src)
```

Comment:
Let's rename these to `gsof_pack_double` in a future PR, and add a comment as to why they're required.

Reply:
https://github.com/ArduPilot/ardupilot/pull/24552